File: leetcode904/leetcode904.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
    public:
        int totalFruit(vector<int>& fruits) {
            int ans = 0, left = 0;
            unordered_map<int, int> cnt;
            for (int right = 0; right < fruits.size(); right++){
                cnt[fruits[right]]++;
                while (cnt.size() > 2)
                {
                    cnt[fruits[left]]--;
                    if (cnt[fruits[left]] == 0){
                        cnt.erase(fruits[left]);
                    }
                    left ++;
                }
                ans = max(ans, right - left + 1);
            }
            return ans;
        }
    };

File: leetcode904/leetcode904.py
from collections import defaultdict


class Solution:
    def totalFruit(self, fruits: List[int]) -> int:
        ans = left = 0
        cnt  = defaultdict(int)
        for right, fruit in enumerate(fruits):
            cnt[fruit] += 1
            while len(cnt) > 2:
                cnt[fruits[left]] -= 1
                if cnt[fruits[left]] == 0:
                    cnt.pop(fruits[left])
                left += 1
            ans = max(ans, right - left + 1)
        return ans